File: src/main/java/net/thedragonteam/armorplus/container/ContainerLavaInfuser.java
package net.thedragonteam.armorplus.container;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.IContainerListener;
import net.minecraft.inventory.Slot;
import net.minecraft.item.ItemStack;
import net.thedragonteam.armorplus.api.lavainfuser.LavaInfuserManager;
import net.thedragonteam.armorplus.api.lavainfuser.SlotLavaInfuserFuel;
import net.thedragonteam.armorplus.api.lavainfuser.SlotLavaInfuserOutput;
import net.thedragonteam.armorplus.container.base.ContainerBase;
import net.thedragonteam.armorplus.tileentity.TileLavaInfuser;

import java.util.stream.IntStream;

import static net.minecraft.item.ItemStack.EMPTY;

/**
 * @author Sokratis Fotkatzikis - TheDragonTeam
 */
public class ContainerLavaInfuser extends ContainerBase {
    private static final int ITEM_BOX = 18;
    private int infusionTime = 0;
    private int totalInfusionTime = 0;
    private int infuserInfusingTime = 0;
    private int currentItemInfusingTime = 0;
    private TileLavaInfuser tile;

    public ContainerLavaInfuser(InventoryPlayer playerInventory, TileLavaInfuser tile) {
        this.tile = tile;
        this.addSlotToContainer(new Slot(tile, 0, 69, 35));
        this.addSlotToContainer(new SlotLavaInfuserFuel(tile, 1, 34, 35));
        this.addSlotToContainer(new SlotLavaInfuserOutput(playerInventory.player, tile, 2, 124, 35));

        for (int height = 0; height < 3; height++) {
            for (int width = 0; width < 9; width++) {
                this.addSlotToContainer(new Slot(playerInventory, width + height * 9 + 9, 8 + width * ITEM_BOX, 84 + height * 18));
            }
        }

        IntStream.range(0, 9).mapToObj(index -> new Slot(playerInventory, index, 8 + index * ITEM_BOX, 142)).forEach(this::addSlotToContainer);
    }

    @Override
    public void addListener(IContainerListener listener) {
        super.addListener(listener);
        listener.sendAllWindowProperties(this, this.tile);
    }

    @Override
    public void detectAndSendChanges() {
        super.detectAndSendChanges();
        this.listeners.forEach(listener -> {
            if (this.infusionTime != this.tile.getField(2))
                listener.sendWindowProperty(this, 2, this.tile.getField(2));
            if (this.infuserInfusingTime != this.tile.getField(0))
                listener.sendWindowProperty(this, 0, this.tile.getField(0));
            if (this.currentItemInfusingTime != this.tile.getField(1))
                listener.sendWindowProperty(this, 1, this.tile.getField(1));
            if (this.totalInfusionTime != this.tile.getField(3))
                listener.sendWindowProperty(this, 3, this.tile.getField(3));
        });

        this.infusionTime = this.tile.getField(2);
        this.infuserInfusingTime = this.tile.getField(0);
        this.currentItemInfusingTime = this.tile.getField(1);
        this.totalInfusionTime = this.tile.getField(3);
    }

    @Override
    public void updateProgressBar(int id, int data) {
        this.tile.setField(id, data);
    }


    @Override
    public ItemStack transferStackInSlot(EntityPlayer playerIn, int index) {
        ItemStack itemstack = EMPTY;
        Slot slot = this.inventorySlots.get(index);


        if (slot != null && slot.getHasStack()) {
            ItemStack slotStack = slot.getStack();
            itemstack = slotStack.copy();

            if (index == 2) {
                if (!this.mergeItemStack(slotStack, 3, 39, true)) {
                    return EMPTY;
                }

                slot.onSlotChange(slotStack, itemstack);
            } else if (index != 1 && index != 0) {
                if (!LavaInfuserManager.getInstance().getInfusingResult(slotStack).isEmpty()) {
                    if (!this.mergeItemStack(slotStack, 0, 1, false)) {
                        return EMPTY;
                    }
                } else if (TileLavaInfuser.isItemFuel(slotStack)) {
                    if (!this.mergeItemStack(slotStack, 1, 2, false)) {
                        return EMPTY;
                    }
                } else if (index < 30) {
                    if (!this.mergeItemStack(slotStack, 30, 39, false)) {
                        return EMPTY;
                    }
                } else if (index < 39 && !this.mergeItemStack(slotStack, 3, 30, false)) {
                    return EMPTY;
                }
            }

            if (slotStack.isEmpty()) {
                slot.putStack(EMPTY);
            } else {
                slot.onSlotChanged();
            }

            if (slotStack.getCount() == itemstack.getCount()) {
                return EMPTY;
            }

            slot.onTake(playerIn, slotStack);
        }

        return itemstack;
    }
}

File: src/main/java/net/thedragonteam/armorplus/container/ContainerLavaInfuser.kt
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package net.thedragonteam.armorplus.container

import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.player.InventoryPlayer
import net.minecraft.inventory.IContainerListener
import net.minecraft.inventory.Slot
import net.minecraft.item.ItemStack
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import net.thedragonteam.armorplus.api.lavainfuser.LavaInfuserManager
import net.thedragonteam.armorplus.api.lavainfuser.SlotLavaInfuserFuel
import net.thedragonteam.armorplus.api.lavainfuser.SlotLavaInfuserOutput
import net.thedragonteam.armorplus.container.base.ContainerBase
import net.thedragonteam.armorplus.tileentity.TileEntityLavaInfuser

class ContainerLavaInfuser(playerInventory: InventoryPlayer, private val tile: TileEntityLavaInfuser) : ContainerBase() {
    private var cookTime: Int = 0
    private var totalCookTime: Int = 0
    private var furnaceBurnTime: Int = 0
    private var currentItemBurnTime: Int = 0

    init {
        this.addSlotToContainer(Slot(tile, 0, 69, 35))
        this.addSlotToContainer(SlotLavaInfuserFuel(tile, 1, 34, 35))
        this.addSlotToContainer(SlotLavaInfuserOutput(playerInventory.player, tile, 2, 124, 35))

        0.rangeTo(2).forEach { i -> 0.rangeTo(8).forEach { j -> this.addSlotToContainer(Slot(playerInventory, j + i * 9 + 9, 8 + j * ITEM_BOX, 84 + i * 18)) } }

        0.rangeTo(8).forEach { k -> this.addSlotToContainer(Slot(playerInventory, k, 8 + k * ITEM_BOX, 142)) }
    }

    override fun addListener(listener: IContainerListener) {
        super.addListener(listener)
        listener.sendAllWindowProperties(this, this.tile)
    }


    /**
     * Looks for changes made in the container, sends them to every listener.
     */
    override fun detectAndSendChanges() {
        super.detectAndSendChanges()

        for (listener in this.listeners) {

            if (this.cookTime != this.tile.getField(2))
                listener.sendWindowProperty(this, 2, this.tile.getField(2))

            if (this.furnaceBurnTime != this.tile.getField(0))
                listener.sendWindowProperty(this, 0, this.tile.getField(0))

            if (this.currentItemBurnTime != this.tile.getField(1))
                listener.sendWindowProperty(this, 1, this.tile.getField(1))

            if (this.totalCookTime != this.tile.getField(3))
                listener.sendWindowProperty(this, 3, this.tile.getField(3))
        }

        this.cookTime = this.tile.getField(2)
        this.furnaceBurnTime = this.tile.getField(0)
        this.currentItemBurnTime = this.tile.getField(1)
        this.totalCookTime = this.tile.getField(3)
    }

    @SideOnly(Side.CLIENT)
    override fun updateProgressBar(id: Int, data: Int) {
        this.tile.setField(id, data)
    }

    /**
     * Take a stack from the specified inventory slot.
     */
    override fun transferStackInSlot(playerIn: EntityPlayer, index: Int): ItemStack {
        var itemstack = ItemStack.EMPTY
        val slot = this.inventorySlots[index]

        if (slot != null && slot.hasStack) {
            val itemstack1 = slot.stack
            itemstack = itemstack1.copy()

            when {
                index == 2 -> {
                    if (!this.mergeItemStack(itemstack1, 3, 39, true)) {
                        return ItemStack.EMPTY
                    }

                    slot.onSlotChange(itemstack1, itemstack)
                }
                index != 1 && index != 0 -> when {
                    !LavaInfuserManager.getInstance().getInfusingResult(itemstack1).isEmpty -> if (!this.mergeItemStack(itemstack1, 0, 1, false)) {
                        return ItemStack.EMPTY
                    }
                    TileEntityLavaInfuser.isItemFuel(itemstack1) -> if (!this.mergeItemStack(itemstack1, 1, 2, false)) {
                        return ItemStack.EMPTY
                    }
                    index in 3..29 -> if (!this.mergeItemStack(itemstack1, 30, 39, false)) {
                        return ItemStack.EMPTY
                    }
                    index in 30..38 && !this.mergeItemStack(itemstack1, 3, 30, false) -> return ItemStack.EMPTY
                }
                !this.mergeItemStack(itemstack1, 3, 39, false) -> return ItemStack.EMPTY
            }

            if (itemstack1.isEmpty) {
                slot.putStack(ItemStack.EMPTY)
            } else {
                slot.onSlotChanged()
            }

            if (itemstack1.count == itemstack.count) {
                return ItemStack.EMPTY
            }

            slot.onTake(playerIn, itemstack1)
        }

        return itemstack
    }

    companion object {
        private val ITEM_BOX = 18
    }
}